### PR TITLE
[SDK-2825] InAppMessages using Flutter Streams

### PR DESCRIPTION
**Pull request by vanessaland**
https://jira.braze.com/browse/SDK-2825

(This change is for both iOS and Android)

# Pull Request/Issue Resolution

- Added a section `Receiving in-app message data` to include information supported by both methods
- Added an additional section for Flutter Streams `Method 1: In-app message data streams (Recommended)`
- Renamed original section to `Method 2: In-app message data callback (Legacy)`

Closes #**[SDK-2825](https://jira.braze.com/browse/SDK-2825)**

#### Is this change associated with a Braze feature/product release?
- [x] Yes - Flutter version 2.6.0
- [ ] No

---

<details>
<summary>✔️ Pull Request Checklist</summary>
<br>

- [x] Check that all links work.
- [x] Ensure you have completed [our Contributors License Agreement](https://www.braze.com/docs/cla/).
- [ ] Tag @Timothy-Kim and @KellieHawks as a reviewer when your work is **done and ready to be reviewed for merge**. Are you an internal product manager? Reference the interal reviewing chart to tag the appropriate reviewer.
- [x] Tag others as reviewers as necessary.
- [x] If you have modified any links, be sure to add redirects to `assets` > `js` > `broken_redirect_list.js`

</details>

<details>
<summary>⭐ Helpful Wiki Shortcuts</summary>
<br>

- [Writing Style Guide](https://docs.google.com/document/d/e/2PACX-1vTluyDFO3ZEV7V6VvhXE4As_hSFwmnFFdU9g6_TrAYTgH1QmbRoEDDdn5GzKAB9vdBbIdyiFdoaJcNk/pub)
- [Image Style Guide](https://docs.google.com/document/d/e/2PACX-1vRJSkwcjmjrTfLDagZccLpOMMyh5NN5SXRZSjz12cRAHbX4OrUmhvCmYpf_p5YB-9r4_jSOQLkicQIH/pub)
- [Styling Test Page](https://www.braze.com/docs/home/styling_test_page/)

</details>

<details>
<summary>❗ ATTN: For PR Reviewers</summary>
<br>

- [ ] Read our [Reviewing a PR page](https://github.com/Appboy/braze-docs/wiki/Reviewing-a-PR) for more on our reviewing suggestions.
- [ ] Read our [Previewing Documentation page](https://github.com/braze-inc/braze-docs/wiki/Previewing-and-Testing-Documentation) to see how to check the deployment.
  - [ ] Preview all changes in the linked Vercel environment by clicking the preview link in the vercel-bot comment in your PR.
</details>

<details>
<summary>❗ ATTN: Internal Reviewing Chart </summary>
<br>
<b>Work at Braze and not sure who to tag for review?</b> <br>Before tagging @timothy-kim or @KellieHawks for a general review, reference the following chart to see if a specific product vertical/reviewer applies to your pull request.
<br><br>
<table>
<tr>
    <td><b>Reviewer</b></td>
    <td><b>Product Vertical</b></td>
  </tr>
  <tr>
    <td>@Timothy-Kim</td>
    <td>Data Infrastructure<br>Platform Infrastructure<br>Quality Infrastructure</td>
  </tr>
  <tr>
    <td>@kelliehawks</td>
    <td>Internal Tools<br>Product Partnerships<br>SDKs<br>SMS</td>
  </tr>
  <tr>
    <td>@bre-fitzgerald</td>
    <td>Channels<br>FIX<br>Intelligence<br>Reporting<br>SMB</td>
  </tr>
  <tr>
    <td>@lydia-xie</td>
    <td>Email (Composition and Infrastructure)<br>Ingestion<br>Messaging</td>
  </tr>
</table>
</details>

### CODE REVIEW

**Comment by vanessaland** [comment]
@cla-bot check

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 1]
```diff
@@ -81,12 +85,29 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
...
 {% endtabs %}
 
-### Replaying the callback for in-app messages
+#### Replaying the callback for in-app messages
 
 To store any in-app messages triggered before the callback is available and replay them once it is set, add the following entry to the `customConfigs` map when intializing the `BrazePlugin`:
 ```dart
 BrazePlugin braze = new BrazePlugin(customConfigs: {replayCallbacksConfigKey: true});
 ```
+### Method 2: In-app message data streams (Preferred)
```
Comment (hokstuff) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:94]: We should move this up as the main (and first) way to integrate receiving IAM data

I also think we can even remove the old callback method and only mention the streams integration to make sure new clients don't integrate with the (soon-to-be deprecated) callback method. I'm curious if others have the same thoughts for that

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 2]
```diff
@@ -81,12 +85,29 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
...
+
+You can set a data stream listener in Dart to receive in-app message data in the Flutter host app.
+
+To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `listen()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when no longer needed.
```
Comment (hokstuff) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:98]:
```suggestion
To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `listen()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when it is no longer needed.
```

**Review by Bucimis — COMMENTED** [review]
Looks good once the ordering is addressed.

Note, we can't merge this PR until Android is complete

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md (hunk 1 above)
Comment (chshapiro) [reply _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:94]: I think we should leave the callback method in the docs for now but streams should definitely go first and we can leave a not by the callback code that that integration is the legacy integration

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 3]
```diff
@@ -56,14 +56,18 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
 {% endtab %}
 {% endtabs %}
 
-## In-app message data callback
+## Receiving in-app message data
+
+To receive in-app message data in your Flutter app, `BrazePlugin` supports 2 methods for receiving in-app message data. The first method uses a data callback and the second method uses [Streams](https://dart.dev/tutorials/language/streams) (Preferred).
```
[the remark this answers is not in the document]
Comment (vanessaland) [reply _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md]: Thanks Jared, I now updated the wording!

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md (hunk 1 above)
Comment (vanessaland) [reply _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md]: Thanks Daniel and Chava! I now updated the page to have the `Method 1: In-app message data streams (Recommended)` first and `Method 2: In-app message data callback (Legacy)` second. I also included an {% alert note %} to mention the legacy method is soon to be deprecated and when migrating to streams, they should remove any callback logic to ensure that an IAM is processed exactly once.

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 4]
```diff
@@ -56,14 +56,38 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
 {% endtab %}
 {% endtabs %}
 
-## In-app message data callback
+## Receiving in-app message data
+
+To receive in-app message data in your Flutter app, `BrazePlugin` supports two methods for receiving in-app message data. The first method uses Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** and the second method uses a data callback **(Legacy)**.
```
Comment (hokstuff) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:61]:
```suggestion
To receive in-app message data in your Flutter app, the `BrazePlugin` supports receiving in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.
```

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 5]
```diff
@@ -56,14 +56,38 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
...
+
+The `BrazeInAppMessage` object supports a subset of fields available in the native model objects, including `uri`, `message`, `header`, `buttons`, `extras`, and more.
+
+{% alert note %} The legacy data callback method is soon to be deprecated. Note that in-app messages can be added to both data streams and data callbacks. If you have already integrated data callbacks and wish to receive only data streams, remove any callback logic to ensure that in-app messages are processed exactly once. {% endalert %}
```
Comment (hokstuff) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:65]:
```suggestion
{% alert note %} The legacy data callback method will soon be deprecated. Note that in-app messages can be added to both data streams and data callbacks. If you have already integrated data callbacks and wish to use data streams, remove any callback logic to ensure that in-app messages are processed exactly once. {% endalert %}
```

**Review by hokstuff — APPROVED** [review]
Small nits, LGTM after

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 6]
```diff
@@ -56,14 +56,38 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
...
+The `BrazeInAppMessage` object supports a subset of fields available in the native model objects, including `uri`, `message`, `header`, `buttons`, `extras`, and more.
+
+{% alert note %} The legacy data callback method will soon be deprecated. Note that in-app messages can be added to both data streams and data callbacks. If you have already integrated data callbacks and wish to use data streams, remove any callback logic to ensure that in-app messages are processed exactly once. {% endalert %}
+
+### Method 1: In-app message data streams (Recommended)
+
+You can set a data stream listener in Dart to receive in-app message data in the Flutter host app.
+
+To begin listening to the stream, use the code below to create a `StreamSubscription` in your Flutter app and call the `subscribeToInAppMessages()` method with a function that takes a `BrazeInAppMessage` instance. Remember to `cancel()` the stream subscription when it is no longer needed.
+```dart
```
Comment (Bucimis) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:72]: add space above

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md [hunk 7]
```diff
@@ -56,14 +56,38 @@ For an example, see [AppDelegate.swift](https://github.com/braze-inc/braze-flutt
 {% endtab %}
 {% endtabs %}
 
-## In-app message data callback
+## Receiving in-app message data
+
+To receive in-app message data in your Flutter app, the `BrazePlugin` supports receiving in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.
```
Comment (Bucimis) [inline _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:61]:
```suggestion
To receive in-app message data in your Flutter app, the `BrazePlugin` supports sending in-app message data using Dart [Streams](https://dart.dev/tutorials/language/streams) **(Recommended)** or by using a data callback **(Legacy)**.
```
Not sure if this is ideal, but imo the original reads a bit repetitive

**Review by Bucimis — APPROVED** [review]

File: _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md (hunk 7 above)
Comment (Bucimis) [reply _docs/_developer_guide/platform_integration_guides/flutter/inapp_messages.md:61]: (same comment on the iOS side)

**Comment by hokstuff** [comment]
@bre-fitzgerald You can merge this PR as well!

**Review by KellieHawks — APPROVED** [review]

**Pull request merged**